test/benchmark.py

```diff
@@ -0,0 +1,263 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+A test suite to check Spanner dialect for SQLAlchemy performance
+in comparison with the original Spanner client.
+"""
+import datetime
+import pprint
+import time
+
+from google.cloud import spanner_dbapi
+from google.cloud.spanner_v1 import Client, KeySet
+from sqlalchemy import (
+    create_engine,
+    insert,
+    select,
+    text,
+    MetaData,
+    Table,
+)
+
+
+def measure_execution_time(function):
+    """Decorator to measure a wrapped method execution time."""
+
+    def wrapper(self, measures):
+        """Execute the wrapped method and measure its execution time.
+
+        Args:
+            measures (dict): Test cases and their execution time.
+        """
+        t_start = time.time()
+        function(self)
+        measures[function.__name__] = round(time.time() - t_start, 2)
+
+    return wrapper
+
+
+class BenchmarkTestBase:
+    """Base class for performance testing.
+
+    Organizes testing data preparation and cleanup.
+    """
+
+    def __init__(self):
+        self._create_table()
+
+        self._one_row = (
+            1,
+            "Pete",
+            "Allison",
+            datetime.datetime(1998, 10, 6).strftime("%Y-%m-%d"),
+            b"123",
+        )
+
+    def _cleanup(self):
+        """Drop the test table."""
+        conn = spanner_dbapi.connect("sqlalchemy-dialect-test", "compliance-test")
+        conn.database.update_ddl(["DROP TABLE Singers"])
+        conn.close()
+
+    def _create_table(self):
+        """Create a table for performace testing."""
+        conn = spanner_dbapi.connect("sqlalchemy-dialect-test", "compliance-test")
+        conn.database.update_ddl(
+            [
+                """
+CREATE TABLE Singers (
+    id INT64,
+    first_name STRING(1024),
+    last_name STRING(1024),
+    birth_date DATE,
+    picture BYTES(1024),
+) PRIMARY KEY (id)
+        """
+            ]
+        ).result(120)
+
+        conn.close()
+
+    def run(self):
+        """Execute every test case."""
+        measures = {}
+        for method in (
+            self.insert_one_row_with_fetch_after,
+            self.read_one_row,
+            self.insert_many_rows,
+            self.select_many_rows,
```

Comment:
I'm curious to know if it's possible to choose between different transaction types and if so, then should we add tests for that. E.g. EF had `SelectMultipleSingersInReadOnlyTransactionEF` and `SelectMultipleSingersInReadWriteTransactionEF`.

Reply:
Also, do any of these tests trigger the checksum computation in dbapi? If not, is it possible to craft a test that does so? E.g. in autocommit mode.

Reply:
You can ignore my first comment. I found the discussion about this in the chat.

Reply:
> Also, do any of these tests trigger the checksum computation in dbapi? If not, is it possible to craft a test that does so? E.g. in autocommit mode.

Checksums are calculated in !autcommit mode, which is the default mode, so tests should already be calculating checksums.